### PR TITLE
[SYCL][FPGA] Update __builtin_intel_fpga_mem

### PR DESCRIPTION
**Pull request by Fznamznon**
According to https://github.com/intel/llvm/pull/4980 this built-in
should take four more constant integer parameters with following default
values:
```
const int32_t AnchorID = -1
const int32_t TargetAnchor = 0
const int32_t Type = 0
const int32_t Cycle = 0
```
The old variant is still allowed for backward compatibility. In case
some parameters are not provided - default value will be emitted in
annotation string.

### CODE REVIEW

**Comment by Fznamznon** [comment]
@shuoniu-intel , please take a look to make sure we have the same understanding of how this is supposed to work. Please also let me know if strings emitted in LLVM IR i.e. `anchor-id, target-anchor, type, cycle` are ok.

**Comment by shuoniu-intel** [comment]
> @shuoniu-intel , please take a look to make sure we have the same understanding of how this is supposed to work. Please also let me know if strings emitted in LLVM IR i.e. `anchor-id, target-anchor, type, cycle` are ok.

LGTM!

**Comment by Fznamznon** [comment]
Ping @elizabethandrews , @premanandrao , @smanna12 on review.

File: clang/test/CodeGenSYCL/intel-fpga-mem-builtin.cpp [hunk 1]
```diff
@@ -65,6 +69,26 @@ void foo(float *A, int *B, State *C, State &D) {
...
+
+  // CHECK-DAG: [[B3:%[0-9]+]] = load i32 addrspace(4)*, i32 addrspace(4)* addrspace(4)* [[Baddr]]
+  // CHECK-DAG: [[PTR11:%[0-9]+]] = call i32 addrspace(4)* @llvm.ptr.annotation{{.*}}[[B3]]{{.*}}[[ANN5]]{{.*}}[[ATT:#[0-9]+]]
+  // CHECK-DAG: store i32 addrspace(4)* [[PTR11]], i32 addrspace(4)* addrspace(4)* %y
+  y = __builtin_intel_fpga_mem(B, PARAM_1 | PARAM_2, 127, 100);
+
+  // CHECK-DAG: [[D1:%[0-9]+]] = load [[STRUCT]] addrspace(4)*, [[STRUCT]] addrspace(4)* addrspace(4)* [[Daddr]]
+  // CHECK-DAG: [[PTR12:%[0-9]+]] = call [[STRUCT]] addrspace(4)* @llvm.ptr.annotation{{.*}}[[D1]]{{.*}}[[ANN6]]{{.*}}[[ATT:#[0-9]+]]
+  // CHECK-DAG: store [[STRUCT]] addrspace(4)* [[PTR12]], [[STRUCT]] addrspace(4)* addrspace(4)* %z
+  z = __builtin_intel_fpga_mem(&D, PARAM_1 | PARAM_2, 128, 4, 5, 6);
```
Comment (premanandrao) [inline clang/test/CodeGenSYCL/intel-fpga-mem-builtin.cpp:91]: Suggestion: perhaps change one of these to use constexpr values?

File: clang/test/CodeGenSYCL/intel-fpga-mem-builtin.cpp [hunk 2]
```diff
@@ -9,8 +9,12 @@ struct State {
   float y;
 };
```
Comment (premanandrao) [inline clang/test/CodeGenSYCL/intel-fpga-mem-builtin.cpp:11]: Please add a comment about what this test is doing.

File: clang/test/SemaSYCL/intel-fpga-mem-builtin.cpp [hunk 3]
```diff
@@ -29,9 +29,7 @@ void foo(float *A, int *B, State *C) {
   x = __builtin_intel_fpga_mem(A, PARAM_1 | PARAM_2, -1);
```
Comment (premanandrao) [inline clang/test/SemaSYCL/intel-fpga-mem-builtin.cpp:29]: Also this test.   Thanks.

File: clang/test/CodeGenSYCL/intel-fpga-mem-builtin.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/intel-fpga-mem-builtin.cpp]: Sure, done.

File: clang/test/CodeGenSYCL/intel-fpga-mem-builtin.cpp (hunk 2 above)
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/intel-fpga-mem-builtin.cpp:14]: Done.

File: clang/test/SemaSYCL/intel-fpga-mem-builtin.cpp (hunk 3 above)
Comment (Fznamznon) [reply clang/test/SemaSYCL/intel-fpga-mem-builtin.cpp:32]: Done.

**Review by smanna12 — DISMISSED** [review]
LGTM

**Review by premanandrao — APPROVED** [review]
LGTM

**Review by premanandrao — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Comment by Fznamznon** [comment]
@shuoniu-intel , Is it safe to merge this one? Does FPGA compiler support new form of the annotations? Or maybe we should wait for the spec?

**Comment by bader** [comment]
> Is it safe to merge this one?

I think the spec must be approved and merged before the implementation.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 4]
```diff
@@ -20449,23 +20449,44 @@ RValue CodeGenFunction::EmitIntelFPGAMemBuiltin(const CallExpr *E) {
...
 
   // Create the pointer annotation
   Function *F =
       CGM.getIntrinsic(llvm::Intrinsic::ptr_annotation, PtrVal->getType());
   SmallString<256> AnnotStr;
   llvm::raw_svector_ostream Out(AnnotStr);
 
-  Optional<llvm::APSInt> Params =
-      E->getArg(1)->getIntegerConstantExpr(getContext());
+  Optional<llvm::APSInt> Params = E->getArg(1)->getIntegerConstantExpr(Ctx);
```
Comment (mlychkov) [inline clang/lib/CodeGen/CGBuiltin.cpp:20460]: May we use AddOptionalArgValue for first two parameters as well (maybe to rename lambda to AddArgValue)?

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 5]
```diff
@@ -20449,23 +20449,44 @@ RValue CodeGenFunction::EmitIntelFPGAMemBuiltin(const CallExpr *E) {
...
+            ? E->getArg(NumOfArg)->getIntegerConstantExpr(Ctx)
+            : APSInt::get(DefaultValue);
+    assert(IntVal.hasValue() && "Constant arg isn't actually constant?");
+    Out << "{" << StringToAdd << ":" << toString(*IntVal, 10) << "}";
+  };
+
+  AddOptionalArgValue(-1, 3, "anchor-id");
+  AddOptionalArgValue(0, 4, "target-anchor");
+  AddOptionalArgValue(0, 5, "type");
+  AddOptionalArgValue(0, 6, "cycle");
```
Comment (mlychkov) [inline clang/lib/CodeGen/CGBuiltin.cpp:20488]: May we use AddOptionalArgValue for first two parameters as well (maybe to rename lambda to AddArgValue)?
```suggestion
  auto AddArgValue = [&E, &Ctx, &Out](unsigned NumOfArg,
                                              StringRef StringToAdd,
                                              int DefaultValue = MIN_INT) {
    Optional<llvm::APSInt> IntVal =
        (E->getNumArgs() > NumOfArg)
            ? E->getArg(NumOfArg)->getIntegerConstantExpr(Ctx)
            : APSInt::get(DefaultValue);
    assert(IntVal.hasValue() && "Constant arg isn't actually constant?");
    Out << "{" << StringToAdd << ":" << toString(*IntVal, 10) << "}";
  };

  AddArgValue(1, "anchor-id");
  AddArgValue(2, "anchor-id");
  AddArgValue(3, "anchor-id", -1);
  AddArgValue(4, "target-anchor", 0);
  AddArgValue(5, "type", 0);
  AddArgValue(6, "cycle", 0);
```

**Comment by shuoniu-intel** [comment]
> @shuoniu-intel , Is it safe to merge this one? Does FPGA compiler support new form of the annotations? Or maybe we should wait for the spec?

It should be safe since your implementation is backwards-compatible. 
I'm okay with it if reviewers think this PR should wait for the spec.

**Comment by Fznamznon** [comment]
> > @shuoniu-intel , Is it safe to merge this one? Does FPGA compiler support new form of the annotations? Or maybe we should wait for the spec?
> 
> It should be safe since your implementation is backwards-compatible. I'm okay with it if reviewers think this PR should wait for the spec.

Please note that even though the new builtin is backward compatible for users in C++ code, this patch adds permanent generation of four new default values in the annotation string in LLVM IR, so after this patch the output is always a bit different. I was wondering if FPGA compiler will be ok with that.
I've ran extended pre commit in order to understand if some of the compiler's components are not ok with this change. Even though the tests have passed, it can be explained by lack of the tests. So I'm double checking with people.
BTW, @MrSidims , do you know if SPIR-V translator will be ok with the new LLVM IR emitted?

**Comment by shuoniu-intel** [comment]
Let this PR wait for the spec then. Meanwhile I'll verify if the FPGA compiler is okay with the new annotation string.

Btw this will be the new look of the annotation string output from clang and SPIR-V right?
`[[ANN1:@.str[\.]*[0-9]*]] = {{.*}}{params:384}{cache-size:0}{anchor-id:-1}{target-anchor:0}{type:0}{cycle:0}`

**Comment by MrSidims** [comment]
> Please note that even though the new builtin is backward compatible for users in C++ code, this patch adds permanent generation of four new default values in the annotation string in LLVM IR, so after this patch the output is always a bit different. I was wondering if FPGA compiler will be ok with that. I've ran extended pre commit in order to understand if some of the compiler's components are not ok with this change. Even though the tests have passed, it can be explained by lack of the tests. So I'm double checking with people. BTW, @MrSidims , do you know if SPIR-V translator will be ok with the new LLVM IR emitted?

> do you know if SPIR-V translator will be ok with the new LLVM IR emitted?

I expect, that it will map LLVM IR construct on `UserSemantic` decoration and translate it backwards without any problems. If it's not happening - then it's a bug in the translator.

>  I've ran extended pre commit in order to understand if some of the compiler's components are not ok with this change

I don't think we have OK test coverage for FPGA h/w here :(

In general, if the compiler doesn't know what to do with annotation information - the annotation intrinsics should be ignored (if seen otherwise - it's a bug in the device's backend IMO).

**Comment by Fznamznon** [comment]
> Btw this will be the new look of the annotation string output from clang and SPIR-V right? `[[ANN1:@.str[\.]*[0-9]*]] = {{.*}}{params:384}{cache-size:0}{anchor-id:-1}{target-anchor:0}{type:0}{cycle:0}`

Yeah, It will look like the old variant but with default values `{anchor-id:-1}{target-anchor:0}{type:0}{cycle:0}` added at the end.
An example of normal emitted IR:
```
@.str = private unnamed_addr constant [75 x i8] c"{params:384}{cache-size:0}{anchor-id:-1}{target-anchor:0}{type:0}{cycle:0}\00", section "llvm.metadata"
```

**Pull request closed without merging**